File: udp-speed-meter/main/rc4.c
/*
 * rc4.c
 *
 *  Created on: Apr 9, 2019
 *      Author: stanislav.perchenko
 */

#include "rc4.h"

static uint8_t S[256+1];
static size_t i, j;

void rc4_init(uint8_t *key, uint8_t key_length) {
	uint8_t temp;
	for(i = 0; i < 256; i++) S[i] = i;
	for(i = j = 0; i < 256; i++) {
		j = (j + key[i % key_length] + S[i]) % 256;
		temp = S[i];
		S[i] = S[j];
		S[j] = temp;
	}
	i = j = 0;
}


void rc4_output(uint8_t *dst, size_t index, size_t length) {
	size_t count, acc;
	uint8_t temp;
	for (count=0; count<length; count ++, index++) {
		i = ( i + 1 ) % 256;
		j = ( j + S[ i ] ) % 256;
		temp = S[ j ];
		S[ j ] = S[ i ];
		S[ i ] = temp;
		acc = S[i] + S[j];
		dst[index] = S[acc % 256];
	}
}

File: app/src/main/java/com/alperez/esp32/netspeed_client/core/RC4.java
package com.alperez.esp32.netspeed_client.core;

/**
 * Created by stanislav.perchenko on 4/10/2019
 */
public class RC4 {
    private final byte S[] = new byte[256];
    private int i, j;


    public RC4() {
        init(new byte[4]);
    }

    public synchronized void init(byte[] key) {
        for (int idx=0; idx<256; idx++) S[idx] = (byte)idx;

        byte temp;
        for (int idxI=0, idxJ=0; idxI < 256; idxI++) {
            idxJ = (idxJ + (key[idxI % key.length] & 0x0000_00FF) + (S[idxI] & 0x0000_00FF)) % 256;
            temp = S[idxI];
            S[idxI] = S[idxJ];
            S[idxJ] = temp;
        }

        i = j = 0;
    }

    public synchronized void nextBytes(byte[] dst, int index, int length) {
        byte temp;
        for (int counter = 0; counter < length; counter++) {

            i = (i + 1) % 256;
            j = (j + (S[i]&0x0000_00FF)) % 256;

            temp = S[j];
            S[j] = S[i];
            S[i] = temp;

            dst[index++ ] = S[ ((S[i]&0x0000_00FF) + (S[j]&0x0000_00FF))  % 256 ];
        }
    }
}
